src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,484 @@
+use super::{add, CellValue, EccConfig, EccPoint, EccScalarVar, Var};
+use crate::{
+    circuit::gadget::utilities::copy,
+    constants::{NUM_COMPLETE_BITS, T_Q},
+};
+use std::ops::{Deref, Range};
+
+use bigint::U256;
+use ff::PrimeField;
+use halo2::{
+    arithmetic::FieldExt,
+    circuit::{Layouter, Region},
+    plonk::{Column, ConstraintSystem, Error, Expression, Fixed, Permutation, Selector},
+    poly::Rotation,
+};
+
+use pasta_curves::pallas;
+
+mod complete;
+mod incomplete;
+mod overflow;
+
+// Bits used in incomplete addition. k_{254} to k_{4} inclusive
+const INCOMPLETE_LEN: usize = pallas::Scalar::NUM_BITS as usize - 1 - NUM_COMPLETE_BITS;
+const INCOMPLETE_RANGE: Range<usize> = 0..INCOMPLETE_LEN;
+
+// Bits k_{254} to k_{4} inclusive are used in incomplete addition.
+// The `hi` half is k_{254} to k_{130} inclusive (length 125 bits).
+const INCOMPLETE_HI_RANGE: Range<usize> = 0..(INCOMPLETE_LEN / 2);
+
+// Bits k_{254} to k_{4} inclusive are used in incomplete addition.
+// The `lo` half is k_{129} to k_{4} inclusive (length 126 bits).
+const INCOMPLETE_LO_RANGE: Range<usize> = (INCOMPLETE_LEN / 2)..INCOMPLETE_LEN;
```

Comment:
Add a static assertion that the boundary is at 130, i.e. that `INCOMPLETE_LO_RANGE.len() + 1 + NUM_COMPLETE_BITS == 130`.

Reply:
Add a link to the book here.